### PR TITLE
Added core::cmp::Reverse for sort_by_key reverse sorting

### PR DESCRIPTION
**Pull request by mitsuhiko**
I'm not sure if this is the best way to go about proposing this feature but it's pretty useful. It allows you to use `sort_by_key` and return tuples where a single item is then reversed to how it normally sorts.

I quite miss something like this in Rust currently though I'm not sure if this is the best way to implement it.

### CODE REVIEW

**Comment by sfackler** [comment]
This seems like a nice ergonomic win to me. Thoughts @rust-lang/libs?

**Comment by ranma42** [comment]
It would be possible to relax the constraints a bit, so that `T` is allowed to only implement partial orderings, like [this](https://play.rust-lang.org/?gist=0d7882c54b48a51975558425c260eb7d&version=nightly&backtrace=0).

**Comment by nagisa** [comment]
Stability attributes should be fixed.

**Comment by malbarbo** [comment]
How about a free function `cmp::reverse`? `Reverse` seems to generic...

**Comment by mitsuhiko** [comment]
@malbarbo what would the function do? Create a "Reverse" object?

**Comment by malbarbo** [comment]
@mitsuhiko Yes. This is similar to free functions in `iter` module.

**Comment by mitsuhiko** [comment]
I was going with `std::num::Wrapping` in mind which is also just a wrapper / newtype over something. Feels wasteful to add a function that does nothing other than making an enum which needs to be public anyways.

**Comment by mitsuhiko** [comment]
Alternative name proposals:

* `std::cmp::Rev`
* `std::cmp::Invert`
* `std::cmp::Inv`

**Comment by alexcrichton** [comment]
Seems like a nice addition to me!

**Comment by alexcrichton** [comment]
I personally like `Reverse` but would be ok with `Invert` as well.

**Comment by scottmcm** [comment]
Good to see it handling `PartialOrd` as well now.

I don't like `Invert`, since that sounds like `1/x`.  I'm not a fan of just `Reverse` either, since `v.reverse()` and `Reverse(v)` would both compile but do _very_ different things.

Maybe `ReverseOrder`?  This doesn't feel like something that needs a particularly short name.

Or maybe something involving `Descending`?  `v.sort_by_key(|k| Descending(k))` reads decently...

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -322,6 +322,40 @@ impl Ordering {
...
+///
+/// ```
+/// use std::cmp::Reverse;
+///
+/// let mut v = vec![1, 2, 3, 4, 5, 6];
+/// v.sort_by_key(|&num| (num > 3, Reverse(num)));
+/// assert_eq!(v, vec![3, 2, 1, 6, 5, 4]);
+/// ```
+#[derive(PartialEq, Eq, Debug)]
+#[stable(feature = "rust1", since = "1.18.0")]
```
Comment (aturon) [inline src/libcore/cmp.rs]: This will need to start out unstable.

**Comment by mitsuhiko** [comment]
In my own project I called it `RevKey` but I found that confusing. I do like the idea of `Descending` which is what some SQL frameworks do. For instance in Python's SQLAlchemy `column` sorts upwards and `column.desc()` sorts backwards.

**Comment by aturon** [comment]
We discussed this PR in the libs triage meeting today, and everyone was in favor of the addition. We're also fine with both the name and API. (It's a slight shift from other APIs in `std`, in that it exposes a constructor for a tuple struct directly, but that is where idioms have headed in general).

Other than changing the annotations to `unstable`, I think this is good to land. r=me after that change.

**Comment by aturon** [comment]
@mitsuhiko Ah sorry, to clarify the unstable attribute will need to point to a fresh "tracking issue" like https://github.com/rust-lang/rust/issues/40494 (make sure to tag B-unstable). You can go ahead and create that.

**Comment by mitsuhiko** [comment]
@aturon i created the issue but lack the rights to flag it appropriately. So it's untagged for now.

**Comment by BurntSushi** [comment]
Thanks @mitsuhiko! :-)

**Comment by mitsuhiko** [comment]
Wohoo \o/ Thanks :)

File: src/libcore/cmp.rs [hunk 2]
```diff
@@ -322,6 +322,41 @@ impl Ordering {
...
+///
+/// ```
+/// #![feature(reverse_cmp_key)]
+/// use std::cmp::Reverse;
+///
+/// let mut v = vec![1, 2, 3, 4, 5, 6];
+/// v.sort_by_key(|&num| (num > 3, Reverse(num)));
+/// assert_eq!(v, vec![3, 2, 1, 6, 5, 4]);
+/// ```
+#[derive(PartialEq, Eq, Debug)]
```
Comment (rkjnsn) [inline src/libcore/cmp.rs]: Why do we derive PartialEq and Eq here, when we are specifying custom implementations below?

Comment (ranma42) [inline src/libcore/cmp.rs]: There is no custom implementation for `PartialEq`, only for `Ord` and `PartialOrd`.

Comment (rkjnsn) [inline src/libcore/cmp.rs]: Right, of course. I don't know where my brain was. Sorry about that.

**Pull request merged**